### DISCUSSION TITLE
Known future

### DISCUSSION THREAD

**Discussion by thierrymarianne** [idno/idno#3109] [General]
Hey! 

Do you folks need a hand with triaging issues, 
merging pull-requests? I've been in love with known since [2014](https://thierry.sioukam.re/2014/10/07/it-feels-good-having-known-idnoidno-installed-at-home-indieweb) and 
i feel genuinely concerned by its roadmap sustainability  
(as the last tag has been pushed two years ago even though 
patches are continuously applied to `dev` branch - and I thank you for that 🙏🏼). 

I could offer some help. 
I'm not sitting on too much free time 
but i could spare some for helping around.

Sorry if I'm already missing other potentially similar discussions.
I'd be glad to be redirected to other existing thread, issue or topics
with regard to contributions.

**Comment by lindner** [comment]
Thanks Marianne!  The known project could use some new contributors.  I'll reach out directly.

**Comment by benwerd** [comment]
+1 to what @lindner said - it really could. I wish I'd been able to dedicate more time to it over the last few years, but I use it every day myself.

**Comment by benwerd** [comment]
I'll reply a little more in-depth here:

@lindner has really been fighting the good fight with Known by himself for a little while. I've had a series of personal challenges that have made it a bit tough to work on it for extended periods: the terminal illness of a parent, and then a new child. All stuff I feel privileged to help with but work like this has certainly fallen by the wayside.

I would love to see us move to a Known 2.0. Right now I think there are a few things that make Known harder to work on:

* We've got some dependencies on external repositories that may or may not be maintained (for, eg, language file generation). Some of that code also seems to be a bit finicky in some situations. I'd like to remove those dependencies again and bring that functionality into core.
* I'm really not sure that we need to require minification for static resources, or webpack. We could argue those points.
* I'm wondering if we should go to a monorepo and bring stock plugins back here rather than having a large number of outside repositories to keep track of.

And then, once this is done, working on modernization that has been discussed elsewhere:

* I think PHP 8 is a perfectly fine target
* Removing (or, frankly, just upgrading in the first instance) Bootstrap
* ActivityPub support
* Some kind of standardization around what actually shows up on a Known website homepage across themes

As @lindner has mentioned elsewhere, there are funds in the OpenCollective to work on these things and other stuff that the community finds useful. I'd love to start allocating those.

**Comment by benwerd** [comment]
Of course, this is an open project. That's my take. What do you all think?

**Reply by thierrymarianne** [reply]
Thank you for leaving the door open to further suggestions!

P.S. Little note aside, about [the reply below](https://github.com/idno/known/discussions/3109#discussioncomment-3715633),
please don't hesitate to tell to re-draft some of my sentences.
Synthesis is not my strongest suit and I'd like to get better at it :)

**Comment by thierrymarianne** [comment]
Hello @lindner, 

Thank you for helping in connecting the dots.
I understand a little bit better the kind of challenges,
you've been facing so far :)

Hello @benwerd, 

I'm sorry to hear about what you had to go through,
and also happy of hearing the news about a new child.
Congratulations to your family members!

[…]

Hard to move forward considering 
what free and open source project maintainers go through with 
all of their parallel lives 🙏🏼 (thank you so much!) 😅

I got to try though…

If I had to rephrase the parts I understood with a third-party perspective :
 - [ ] to identify critical dependencies 
   eg language file generation
 - [ ] to discuss with the Known ecosystem plugin maintainers 
   about dependencies maintenance
 - [ ] to discuss deployment tasks simplification like assets minification
   by considering removing tooling or replacing tooling (such as webpack)
 - [ ] to discuss moving to a monorepo architecture including stock plugins,
   which may simplify maintenance with a constellation of gravitating repositories

I'd leave the second part as is for later since 
it clearly sound secondary apart perhaps for PHP 8.1 support as  
PHP 8.0 active support will be over by the end of [Nov 2021](https://www.php.net/supported-versions)     
  and there will fortunately still be security support until next year (by the end of Nov 2023) 😌.

I'd suggest we _refine_ before _prioritization_ those items
per your _convenience_, and as our _respective schedules_ allow it
with other _willing community members_.

Being in the process of becoming a student again (with a family of my own, myself),
I wouldn't be able to work full-time on the project but 
I could move from a small deliverable to the next given that
we'd figure how to agree on satisfying enough terms for us all. 😊 

I wonder how messages are usually voiced
throughout the community when it comes to preparing or making decisions collectively.

Thank you very much for taking the time of this discussion.

**Comment by lindner** [comment]
Greetings all!

Thanks @benwerd for getting us all to this point.  And thank you @thierrymarianne for getting involved.

This is an open source project so you can really just start on whatever inspires you.   There are a lot of open issues that could be attended to.  I could try to find you a starter issue to try to fix if that'd help.  Can you give us an idea of your experience level with PHP, javascript, css, and the various tooling used to build web sites?  That would help.

Have you had a chance to build and run Known from a direct git checkout?  Can you get tests running.  All of these are documented but take a little setup time.

For example a super easy one would be https://github.com/idno/known/issues/2540 -- then you could get your workflow / pull requests tested out.

Then progressively more complicated stuff can happen and then we can have you start maintaining the github side of things and so on...

**Reply by thierrymarianne** [reply]
Hey @lindner,

I've got fresh development Known instances 🎉.

I've tried running all testing commands available in the [documentation](https://docs.withknown.com/en/latest/developers/testing/).

All results have been reported in this [pull-request](https://github.com/thierrymarianne/contrib-known/pull/1)

<details>

<summary>Click to see some screenshots of working instances</summary>

Both CLI installer and Web installers have been used successfully  
to set up separate instances with
 - a MySQL database - based on [official mysql:8-debian container image](https://hub.docker.com/layers/library/mysql/8-debian/images/sha256-12477b35d174c1a410da24a05c4884693836e2a863521aa339a1298ce48b8369?context=explore):
 - a PHP 8.1.11 runtime - based on [official php:8.1-fpm-bullsey container image](https://hub.docker.com/layers/library/php/8.1-fpm-bullseye/images/sha256-798fdbaaf776a0f27a1c3988602630e2cb4d03450acefce229e91a1e5fa1d4f4?context=explore) 

![2022-10-10_php-8 1 11_cli-installation](https://user-images.githubusercontent.com/1053622/194946257-26cc9c53-659b-45c8-9ab1-5bcfa8b5e99b.png)

![2022-10-10_08-19-13_web-installer](https://user-images.githubusercontent.com/1053622/194946232-6013e941-a979-4229-a562-b4ee4f8ad6ae.png)

`known.php` script appears to be working as expected as far as I could tell:

![2022-10-10_php-8 1 11_running-command-line](https://user-images.githubusercontent.com/1053622/194947098-5a45cf55-9236-412b-992d-a7d768aed962.png)
</details>

When running unit test, there is one failing test even though it appears to be related to an arbitrary domain used in HSTS test (not sure about what is the intent).

**Reply by thierrymarianne** [reply]
#2540 has been taken care of by https://github.com/idno/known/pull/3131 which branch cumulates - 
 - dependencies updates suggested by dependabot
 - phpcs, phpcbf runs
 - the actual EditorConfig file addition

**Comment by thierrymarianne** [comment]
Hello!

Glad to read you again and to be relaxed with regard
to how things are baked around 😌

I've had some interesting experiences (to say the least)
with PHP-based applications along my career (which started circa 2009),
along with fundamentals languages like CSS, JavaScript and HTML.

<details>

<summary>Click for more details</summary>

I can't say that I don't have a hard time keeping up to date
with most of it today, but I sort of manage.

At the moment, you could have an idea of what I could do
or better said, what I allow myself to do when nobody is looking at it 🙈
on some of my spare time by visiting [revuedepresse repositories](https://https://github.com/orgs/revuedepresse/repositories),
especially this [API and its two routes](https://github.com/revuedepresse/org.revue-de-presse.api/blob/5b1e4bc429831b1ffee8335c451627b54814c11c/config/routes/api.yaml) looking a bit scary  
and only reflecting years of experiments on the side with various [components](https://speakerdeck.com/thierrymarianne/comment-entretenir-un-vieux-cyborg-a-composants-symfony).
It blew out of proportions and yet, I'd rather keep most of the code as is as a reminder,
which may not be subject to forming part of an attack surface. 🤞🏼
There is also a [command-line application](https://github.com/revuedepresse/org.revue-de-presse.api/tree/f3dc04420344f95e22722482b141cab2ac8c9519) in there somewhere.

In the past and within [SMEs](https://en.wikipedia.org/wiki/Small_and_medium-sized_enterprises) (adding the link as a reminder of this translation to myself),
I've also worked with former colleagues with command-query separation  
in event-sourcing oriented application, especially with regard to payment systems.
</details>

Over the years, I got used to re-deploying regularly [Known instances](https://sioukam.re/)  
to production by cloning a fresh `idno/known` repository,
before removing the unnecessary files before recycling related web services.

I've never run the tests from a development checkout yet  
and I'll let you know as soon as I've got something running in a local environment,  
since I've never really had to implement anything apart from 
 - https://github.com/idno/known/pull/2876 (which required external dependencies modifications) and 
 - https://github.com/idno/known/pull/3108 (I didn't know how to test, nor how to evolve existing tests).

Thank you for pointing out how https://github.com/idno/known/issues/2540 may belong the low-hanging fruits!   

I'm going with that one and I will let you know how it's going within the next couple of weeks.

**Comment by lindner** [comment]
Ahaha, my apologies for not remembering your previous contributions!

Tests are documented at https://github.com/idno/known/blob/dev/docs/developers/testing.md  Improvements welcome!

Definitely let's move forward with expanding your impact.

**Reply by thierrymarianne** [reply]
Ha ha, everything is fine.   
I didn't expect you to carry the memory of 1150+ issues, pull-requests with you. 😉

Thank you for the additional pointer!
